lib/dns/verifier.js

```diff
@@ -0,0 +1,162 @@
+require('es6-promise').polyfill();
+require('isomorphic-fetch');
+const packet = require('dns-packet');
+const SingedSet = require('./signed_set.js');
+const Util = require('../util.js');
+const getHeader = Util.getHeader;
+const getKeyTag = Util.getKeyTag;
+const checkDigest = Util.checkDigest;
+
+var TRUST_ANCHORS = [
+  {
+    name: ".",
+    type: "DS",
+    class: "IN",
+    data:{
+      keyTag: 19036,
+      algorithm: 8,
+      digestType: 2,
+      digest: new Buffer("49AAC11D7B6F6446702E54A1607371607A1A41855200FD2CE1CDDE32F24E8FB5", "hex")
+    }
+  },
+  {
+    name: ".",
+    type: "DS",
+    class: "IN",
+    data:{
+      keyTag: 20326,
+      algorithm: 8,
+      digestType: 2,      
+      digest: new Buffer("E06D44B80B8F1D39A95C0B0D7C65D08458E880409BBC683457104237C7F8EC8D", "hex")
+    }
+  },
+  // This is dummy entry for testing
+  {
+    name: ".",
+    type: "DS",
+    class: "IN",
+    data:{
+      keyTag: 5647,
+      algorithm: 253,
+      digestType: 253,      
+      digest: new Buffer([])
+    }
+  }
+]
+
+async function query(qtype, name){
+  let buf = packet.encode({
+    type: 'query',
+    id: 1,
+    flags: packet.RECURSION_DESIRED,
+    questions: [{
+      type: qtype,
+      class: 'IN',
+      name: name
+    }],
+    additionals: [{
+      type: 'OPT',
+      name: '.',
+      udpPayloadSize: 4096,
+      flags: packet.DNSSEC_OK
+    }]
+  })
+  return  await getDNS(buf);
+}
+    
+async function queryWithProof(qtype, name){
+  let r = await query(qtype, name);
+  let sigs = await filterRRs(r.answers, 'RRSIG');
+  let rrs = await getRRset(r.answers, name, qtype);
+  let ret;
+  for(const sig of sigs){
+    ret = await verifyRRSet(sig, rrs);
+    if(ret){
+      ret.push(new SingedSet(name, sig, rrs));
+      return ret;
+    }
+  }
+}
+
+async function verifyRRSet(sig, rrs) {
+  let sigHeaderName = sig.name;
+  let rrsHeaderRtype = rrs[0].type;
+  let sigdata = sig.data;
+  let rrsdata = rrs[0].data[0];
+  let keys = [];
+  let sets;
+  let signersName = sigdata.signersName;
+
+  if(sigHeaderName == sigdata.signersName && rrsHeaderRtype == 'DNSKEY') {
+    keys = rrs;
+	}else{
+    // Find the keys that signed this RRSET
+    sets = await queryWithProof('DNSKEY', sigdata.signersName);
+    if(sets){
```

Comment:
Is there any point continuing if no sets were returned here?

Reply:
Fixed